src/repr/scalar/jsonb.rs

```diff
@@ -7,147 +7,512 @@
-                    }
-                }
-                Value::Object(vals) => {
-                    for (_, v) in vals {
-                        validate(v)?;
-                    }
-                }
-            }
-            Ok(())
-        }
-        validate(&inner)?;
-        Ok(Jsonb { inner })
-    }
-
-    #[allow(clippy::float_cmp)]
-    pub fn from_datum(datum: Datum) -> Self {
-        fn inner(datum: Datum) -> serde_json::Value {
-            use serde_json::Value;
-            match datum {
-                Datum::JsonNull => Value::Null,
-                Datum::True => Value::Bool(true),
-                Datum::False => Value::Bool(false),
-                Datum::Float64(f) => {
-                    let f: f64 = f.into();
-                    if let Some(n) = serde_json::Number::from_f64(f) {
-                        Value::Number(n)
-                    } else {
-                        // This should only be reachable for NaN/Infinity, which
-                        // aren't allowed to be cast to Jsonb.
-                        panic!("Not a valid json number: {}", f)
-                    }
-                }
-                Datum::String(s) => Value::String(s.to_owned()),
-                Datum::List(list) => Value::Array(list.iter().map(inner).collect()),
-                Datum::Dict(dict) => {
-                    Value::Object(dict.iter().map(|(k, v)| (k.to_owned(), inner(v))).collect())
-                }
-                _ => panic!("Not a json-compatible datum: {:?}", datum),
-            }
+    /// Constructs a new `Jsonb` from a [`serde_json::Value`].
+    ///
+    /// Errors if any of the contained integers cannot be represented exactly as
+    /// an [`f64`].
+    pub fn from_serde_json(val: serde_json::Value) -> Result<Self, failure::Error> {
+        let packer = JsonbPacker::new(RowPacker::new()).pack_serde_json(val)?;
+        Ok(Jsonb {
+            row: packer.finish(),
+        })
+    }
+
+    /// Parses a `Jsonb` from a byte slice `buf`.
+    ///
+    /// Errors if the slice is not valid JSON or if any of the contained
+    /// integers cannot be represented exactly as an [`f64`].
+    pub fn from_slice(buf: &[u8]) -> Result<Jsonb, failure::Error> {
+        let packer = JsonbPacker::new(RowPacker::new()).pack_slice(buf)?;
+        Ok(Jsonb {
+            row: packer.finish(),
+        })
+    }
+
+    /// Constructs a [`JsonbRef`] that references the JSON in this `Jsonb`.
+    pub fn as_ref(&self) -> JsonbRef {
+        JsonbRef {
+            datum: self.row.unpack_first(),
         }
+    }
+
+    /// Consumes this `Jsonb` and returns the underlying [`Row`].
+    pub fn into_row(self) -> Row {
+        self.row
+    }
+}
+
+impl FromStr for Jsonb {
+    type Err = failure::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let packer = JsonbPacker::new(RowPacker::new()).pack_str(s)?;
+        Ok(Jsonb {
+            row: packer.finish(),
+        })
+    }
```

Comment:
This probably ends up with one more allocation than we truly desire (`RowPacker::new()` instead of re-using an existing `RowPacker`). We can't change the trait, but ideally we don't end up too reliant on it.

Reply:
👍